### ISSUE TITLE
Leaky streams

### ISSUE THREAD

**Issue by siddharthlatest** [appbaseio/appbase-js#3]
Usage of global request obj inside stream's stop() has leaky side effects.

## Behavior Observation:

I believe this is what is happening (it's my guess based on reading parts of appbase.js; I can be wrong on the specifics, but the problem exists in the lib):

appbase ref object has a field called body, a property of the appbase object which is updated every time the ``streamingRequest()`` method is called. This is where it gets interesting -

I have a response object from a ``streamingRequest()`` method, I call a ``stop()`` on this and in the next execution step call another ``streamingRequest()`` [call it 2nd step] method. If my stop() execution is not fast enough, there’s a chance it will use the body field set in the 2nd step.

## Reproducing:

Reproducing this is hard, it has to be scripted. The margin of error here is in the order of **2ms**. I am testing this with the https://github.com/appbaseio/meetuprsvp-ui code.

Adding two console logs inside the appbase.js ws's ``stop()`` code at [L12360] (https://github.com/appbaseio/appbase-js/blob/using_websocket/browser/appbase.js#L12360) (pasted below):

```js
wsRequest.prototype.stop = function stop() {
	console.log("appbase.js", this.request);
	this.client.ws.removeListener('close', this.closeHandler)
	this.client.ws.removeListener('error', this.errorHandler)
	this.client.ws.removeListener('message', this.messageHandler)
	if(this.resultStream.readable) {
		this.resultStream.push(null)
	}
	var unsubRequest = {}
	for(var key in this.request) {
		unsubRequest[key] = this.request[key]
	}
	unsubRequest.unsubscribe = true
	console.log("appbase.js - closing ws request", unsubRequest)
	if(this.unsubscribed !== true) {
		this.client.ws.send(unsubRequest)
	}
	this.unsubscribed = true
}
```

``this.request`` has different responses in the two console.log() statements; by the time the code execution reaches to the second console.log() - a new ``streamingRequest()`` is called which modifies the ``this.request`` property.

## Possible Fix:

Keeps everything immutable and involves minimal changes to your existing structure

* make request object a part of the returned response stream; and when stop() is called on the stream, use the local request object (instead of the global mutable one) to send the unsubscribe request. 

``Note``: This also has one another nice benefit: as a user, i can debug the request params easily since it's now a part of the responsestream object

**Comment by sacheendra** [comment]
There is no body property on appbase ref. All our objects are immutable. 
This is the code snippet i am using to test. It does not give problems. Can you suggest a snippet to reproduce the problem. 

```js
var streamingClient = new Appbase({
  url: 'https://RIvfxo1u1:dee8ee52-8b75-4b5b-be4f-9df3c364f59f@scalr.api.appbase.io',
  appname: 'createnewtestapp01'
})
var response = streamingClient.streamSearch({
	type: 'books',
	body: {
		query: {
			match_all: {}
		}
	}
})
response.stop()
response = streamingClient.streamSearch({
	type: 'books',
	body: {
		query: {
			match_all: {}
		}
	}
})
response.on('data', console.log.bind(window.console))
setTimeout(function() {
	streamingClient.index({
	  type: 'books',
	  id: '1',
	  body: {"foo": "bar"}
	})
}, 3000)
```

**Comment by siddharthlatest** [comment]
fixed in 47681acfa44e23036352d8b6679213b0d736cfa9.
